ext/opcache/jit/zend_jit_ir.c

```diff
@@ -9829,6 +9829,49 @@ static int zend_jit_do_fcall(zend_jit_ctx *jit, const zend_op *opline, const zen
 	 && opline->opcode != ZEND_DO_ICALL) {
 		bool recursive_call_through_jmp = 0;
 
+		if (opline->opcode == ZEND_DO_FCALL_BY_NAME) {
+			if (!func) {
+				if (trace) {
```

Comment:
I don't understand why we don't need this branch for `ZEND_DO_FCALL`. I don't get any failures with the previous version (just adding `|| opline->opcode == ZEND_DO_FCALL_BY_NAME` to line 9790) and removing this block entirely. Can you clarify?

Reply:
For reference, this changed in 4012d268a64d4d659d974608e5bfea0ccfb62ead.

The symptoms were that for internal functions the deprecation message was emitted twice, as can be seen in this CI run: https://github.com/php/php-src/actions/runs/9449623113/job/26026345712

My understanding is that the deprecation for ZEND_DO_FCALL_BY_NAME is already handled here for internal functions:

https://github.com/php/php-src/blob/11914de72cab4caf5f62896328cac33f89dab77a/ext/opcache/jit/zend_jit_ir.c#L10143-L10184

And I just copied the entire block into the handler for userland functions, but without understanding what it does and thus unable to refactor this in a smarter way :sweat_smile:

Reply:
Ok. So, I pushed a new commit that removes the redundant code. Basically, previously, the JIT had an optimization that would move the deprecation check inside a `func->type == ZEND_INTERNAL_FUNCTION` branch that no longer applies. Now, we can simply use the same check in all cases. Please check if this looks correct to you. If it does, please ask Dmitry to confirm as well.

Reply:
The change definitely makes sense to me.

@dstogov Can you please confirm that the JIT changes look good to you?